src/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -172,11 +172,14 @@ public enum JsonValueType : byte
         False = (byte)6,
         Null = (byte)7,
     }
-    public partial struct JsonWriterOptions
+    public partial struct JsonWriterOptions : IEquatable<JsonWriterOptions>
```

Comment:
Since this is a public surface change, it would need to go through API review. If possible, could you file an issue to make `JsonWriterOptions` `IEquatable`: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md

The primary motivation to make it `IEquatable` was for used within the serializer to support reusability of the writer (i.e. this PR). If that doesn't provide enough value to justify the complexity, then I am not sure if it still makes sense to make it `IEquatable`. We would have to figure out what are the semantics of `Equals` when we add `JavascriptEncoder` as a property on options. Is it reference equality only (imo, yes), or would we say an option is equal if the encoder produces the same string (even if it happens to be a different instance).

Reply:
@ahson was the original intent of the issue you logged to re-use within a given call stack or attempt to share globally?

Reply:
@ahsonkhan  was the original intent of the issue you logged to re-use within a given call stack or attempt to share globally?

Reply:
Whichever one that's feasible, but I was thinking globally, mainly to avoid the re-sizing/copy of the underlying array (and less about saving a few bytes in terms of allocation).

Reply:
Should we continue if there is a small benefit from caching?

Reply:
Looking at the perf results you shared, I don't think it's worth it.

Reply:
Yeah, I agree

Reply:
We are saving a single alloc of the writer and corresponding GC, plus the pooled buffer allocation that the writer has, correct? In exchange for an IEquateable + ConcurrentDictionary with likely need to remove and potentially re-add the writer after each use to support multi-threading. ConcurrentDictionary is very slow to add. IMO we chalk this up as learning\experiment.

Another option that I considered (but am not recommending) is to store a single cached instance of the writer on the JsonSerializerOptions class without a dictionary. If it is "in use" on another thread (using locks) then we just new up a new writer for that case. This would be fast without a dictionary, but may not be expected either and is the reason I don't recommend it - the cached instance would keep a potentially large buffer (a "high-water" mark since it will be the largest buffer encountered so far) around until the options class is GC'd which may never happen process exit since it will likely be stored in a static.

Reply:
> We are saving a single alloc of the writer and corresponding GC, plus the pooled buffer allocation that the writer has, correct?

Also the re-size/copy data. Essentially, never hitting this code path:
https://github.com/dotnet/corefx/blob/d8797a0f1616c1db1e5294964f2826e051f2615e/src/System.Text.Json/src/System/Text/Json/Serialization/PooledBufferWriter.cs#L126-L142

>  IMO we chalk this up as learning\experiment.

Agreed. I still want to experiment with really large payloads and see if we can get any noticeable benefit, even in the best case.

Reply:
Thanks you all for review, guys